### PR TITLE
bpo-39259: smtp.SMTP/SMTP_SSL now reject timeout = 0

### PR DESCRIPTION
**Pull request by corona10**
https://bugs.python.org/issue39259

### CODE REVIEW

**Comment by corona10** [comment]
@vstinner 
Discussion needed:

During I work on this PR, I've noticed that LMTP does not support the timeout parameter.
See: https://docs.python.org/3.9/library/smtplib.html#smtplib.LMTP

However, since LMTP can use the socket created from SMTP
https://github.com/python/cpython/blob/100fafcf20e8fc67cd8ef512074f9c0a253cb427/Lib/smtplib.py#L1076
Also AFAIK, Unix Socket also able to set timeout
```
>>> import socket
>>> sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
>>> sock.settimeout(30)
>>> sock.gettimeout()
30.0
>>> b = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
>>> b.gettimeout()
>>>
```

So in conlcusion, IMHO LMTP needs to support timeout parameter.
If you agree, I would like to create an issue for this.

**Comment by vstinner** [comment]
> IMHO LMTP needs to support the timeout parameter.

If you think so, please open a separated issue. It's not directly related to https://bugs.python.org/issue39259

File: Lib/smtplib.py [hunk 1]
```diff
@@ -1030,13 +1032,12 @@ def __init__(self, host='', port=0, local_hostname=None,
...
             self.context = context
             SMTP.__init__(self, host, port, local_hostname, timeout,
-                    source_address)
+                          source_address)
 
         def _get_socket(self, host, port, timeout):
             if self.debuglevel > 0:
                 self._print_debug('connect:', (host, port))
-            new_socket = socket.create_connection((host, port), timeout,
-                    self.source_address)
```
Comment (vstinner) [inline Lib/smtplib.py:1039]: Wait, where is self.source_address gone? And why tests didn't catch that self.source_address is lost?

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Lib/smtplib.py (hunk 1 above)
Comment (corona10) [reply Lib/smtplib.py:1039]: @vstinner 
self.source_address was not passed as a parameter, do you want to pass it as a parameter?
https://github.com/python/cpython/blob/29cd207bee95b6598250e28cdf133ba4a068008a/Lib/smtplib.py#L310-L311

**Review by corona10 — COMMENTED** [review]
I have made the requested changes; please review again

File: Lib/smtplib.py (hunk 1 above)
Comment (vstinner) [reply Lib/smtplib.py:1039]: Oh wait, you replaced create_connecton() with _get_socket(). I misread the code. It is correct.

**Pull request merged**